### PR TITLE
Feature gs api informer

### LINKED ISSUE #61: scheduler should be able to watch "binding/unbinding of a POD object to this scheduler" event from the API server
1、scheduler should be able to watch "binding/unbinding of a POD object to this scheduler" event from the API server

### PR DESCRIPTION
**Pull request by nkwangjun**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #61

https://github.com/futurewei-cloud/global-resource-scheduler/pull/81#pullrequestreview-563974422

### CODE REVIEW

**Comment by nkwangjun** [comment]
@cathyhongzhang I have already use the same concept for the scheduler. There is no NODE, NODE_INFO, NODES_INFO，all changed to Site， witch is associated with logical attribute availability-zone. And also， there is no method associated with node， for example addnode，nodescore and somethings else. I think it looks clearer and simpler.

File: conf/conf.yaml [hunk 1]
```diff
@@ -2,8 +2,12 @@
 address: 0.0.0.0
 port: 13443
 
+#api server info
+master: 127.0.0.1:8080
+kubeconfig: /var/run/kubernetes/admin.kubeconfig
+
 scheduler_policy: edgecloud
-scheduler_policy_file: /root/work/src/global-resource-scheduler/conf/edgecloud_policy.yaml
```
Comment (cathyhongzhang) [inline conf/conf.yaml:6]: I think it is good to put all global scheduler config under the same conf dir.

File: conf/site_resources.json [hunk 2]
```diff
@@ -1,13 +1,10 @@
 {
-    "site_nodes": [{
-            "edge_site_id": "region1|az1",
+    "site_resources": [{
+            "site_id": "region1|az1",
             "region": "region1",
             "availability_zone": "az1",
-            "operator": "opera1",
```
Comment (cathyhongzhang) [inline conf/site_resources.json:6]: you have availability_zone at site level and then at each host level. Do you really need both, whcih requires the code to do cross checking? The site level availablity_zone info can be derived from the hosts, right?

**Comment by cathyhongzhang** [comment]
> @cathyhongzhang I have already use the same concept for the scheduler. There is no NODE, NODE_INFO, NODES_INFO，all changed to Site， witch is associated with logical attribute availability-zone. And also， there is no method associated with node， for example addnode，nodescore and somethings else. I think it looks clearer and simpler.

Wangjun, thank you very much for doing this clean up. I think it looks much better now

**Review by CoderKevinZhang — APPROVED** [review]
LGTM

File: conf/edgecloud_policy.yaml [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
 - name: EipAvailability
-- name: ExclusiveNode
+- name: ExclusiveSite
 priorities:
 - name: StackAffinity
   weight: 1
 - name: Flavor
   weight: 1
-- name: NodeResourcesLeastAllocated
+- name: SiteResourcesLeastAllocated
```
Comment (cathyhongzhang) [inline conf/edgecloud_policy.yaml:14]: The predicates and priorities defined in this file determines the filter and score plugins used by the scheudling algorithm, right?

File: conf/site_resources.json [hunk 4]
```diff
@@ -1,13 +1,10 @@
...
-            "status": "active",
-            "max_count": 1000,
-            "node_infos": [{
-                "edge_node_id": "edge_node_id",
+            "hosts": [{
+                "host_id": "host_id",
                 "host_name": "name1",
                 "region": "region1",
                 "availability_zone": "az1",
```
Comment (cathyhongzhang) [inline conf/site_resources.json:10]: Why doesn't the site/hosts have geolocation info?

Comment (cathyhongzhang) [reply conf/site_resources.json:10]: With the new terminology, the site maps to a cluster and each site is composed of many nodes, right?

File: docs/design-proposals/global-scheduler/GlobalSchedulerDemo.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  },
+  "status": "Success",
+  "code": 201
+}
+
+# check the schedule result
+root@arktos-wj-dev:~/yaml# kubectl get pods
+NAME       HASHKEY               READY   STATUS   RESTARTS   AGE
+demovm01   8970278610605413590   0/1     Binded   0          4m47s
```
Comment (jshaofuturewei) [inline docs/design-proposals/global-scheduler/GlobalSchedulerDemo.md:85]: Please change the "Binded" to "Bound"

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 6]
```diff
@@ -381,7 +381,7 @@ const (
 func NewControllerInitializers(loopMode ControllerLoopMode) map[string]InitFunc {
 	controllers := map[string]InitFunc{}
 	controllers["endpoint"] = startEndpointController
-	controllers["podgc"] = startPodGCController
+	//controllers["podgc"] = startPodGCController
```
Comment (jshaofuturewei) [inline cmd/kube-controller-manager/app/controllermanager.go:388]: We do not need to disable PodGCController once we applied the change #145. The gc controller will check pod status and put pods whose status are "assigned" or "bound" into terminated pod list for recycling.

File: hack/lib/common.sh [hunk 7]
```diff
@@ -476,6 +476,13 @@ function kube::common::start_kubescheduler {
     SCHEDULER_PID=$!
 }
 
+function kube::common::start_gs_scheduler {
+    GS_SCHEDULER_LOG=${LOG_DIR}/gs-scheduler.log
+    ${CONTROLPLANE_SUDO} "${GO_OUT}/gs-scheduler" > "${GS_SCHEDULER_LOG}" 2>&1 &
```
Comment (jshaofuturewei) [inline hack/lib/common.sh]: Do we need to add scheduler name to start a gs-scheduler?

File: conf/conf.yaml (hunk 1 above)
Comment (nkwangjun) [reply conf/conf.yaml:6]: Yes， I will modify it to relative directory. Thanks

File: conf/site_resources.json (hunk 2 above)
Comment (nkwangjun) [reply conf/site_resources.json:6]: Yes, here it is redundant field, and can be optimized. I will remove this, thanks.

File: hack/lib/common.sh (hunk 7 above)
Comment (nkwangjun) [reply hack/lib/common.sh]: The code you reviewed is outdated.

function kube::common::start_gs_scheduler {
    GS_SCHEDULER_LOG=${LOG_DIR}/gs-scheduler.log
    ${CONTROLPLANE_SUDO} "${GO_OUT}/gs-scheduler" --schedulername="scheduler1"  > "${GS_SCHEDULER_LOG}" 2>&1 &
    GS_SCHEDULER_PID=$!
}

File: conf/site_resources.json (hunk 2 above)
Comment (nkwangjun) [reply conf/site_resources.json:6]: Yes, this can be optimized, I will modify it, thanks

File: conf/edgecloud_policy.yaml (hunk 3 above)
Comment (nkwangjun) [reply conf/edgecloud_policy.yaml:14]: Yes

File: conf/site_resources.json (hunk 4 above)
Comment (nkwangjun) [reply conf/site_resources.json:10]: 1, geolocation info is get from site_infos, different sources from the resource info
2, yes

File: docs/design-proposals/global-scheduler/GlobalSchedulerDemo.md (hunk 5 above)
Comment (nkwangjun) [reply docs/design-proposals/global-scheduler/GlobalSchedulerDemo.md:85]: OK

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 6 above)
Comment (nkwangjun) [reply cmd/kube-controller-manager/app/controllermanager.go:388]: Ok

File: hack/lib/common.sh (hunk 7 above)
Comment (jshaofuturewei) [reply hack/lib/common.sh]: Got it. Is that possible to pass a param like $1 for schedulername?  Kevin might use the shell script to start schedulers by passing scheduler name. Thanks!

**Comment by cathyhongzhang** [comment]
@nkwangjun Thanks for your response on our comments. Could you upload new PR which addresses our comments so that we can merge it? Please run "make update" and those CI tests locally before updating the PR in the github. Thanks!

File: globalscheduler/pkg/scheduler/factory/factory.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+// NewPodInformer creates a shared index informer that returns only non-terminal pods.
+func NewPodInformer(client clientset.Interface, resyncPeriod time.Duration) coreinformers.PodInformer {
+	selector := fields.ParseSelectorOrDie(
+		"status.phase!=" + string(v1.PodSucceeded) +
+			",status.phase!=" + string(v1.PodFailed))
+	lw := cache.NewListWatchFromClient(client.CoreV1(), string(v1.ResourcePods), metav1.NamespaceAll, selector)
+	return &podInformer{
+		informer: cache.NewSharedIndexInformer(lw, &v1.Pod{}, resyncPeriod, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}),
+	}
```
[the remark this answers is not in the document]
Comment (nkwangjun) [reply globalscheduler/pkg/scheduler/factory/factory.go:52]: I have already do this as you said, and I have already tested it. Please check the eventhandler code please.

**Comment by nkwangjun** [comment]
> @nkwangjun Thanks for your response on our comments. Could you upload new PR which addresses our comments so that we can merge it? Please run "make update" and those CI tests locally before updating the PR in the github. Thanks!
I have already run the "make update" and those CI tests locally before updating

File: globalscheduler/pkg/scheduler/scheduler.go [hunk 9]
```diff
@@ -476,36 +492,53 @@ func (sched *Scheduler) buildFramework() error {
...
+	}
+
+	sched.stackQueue = internalqueue.NewSchedulingQueue(stopCh, sched.SchedFrame)
+	sched.InformerFactory = informers.NewSharedInformerFactory(client, 0)
+	sched.PodInformer = factory.NewPodInformer(client, 0)
+	return nil
+}
+
 // Init resource cache informer
 func InitSchedulerInformer(stopCh <-chan struct{}) {
```
[the remark this answers is not in the document]
Comment (nkwangjun) [reply globalscheduler/pkg/scheduler/scheduler.go]: No

**Comment by nkwangjun** [comment]
This pr will be closed， and it is replaced by  #146

File: hack/lib/common.sh (hunk 7 above)
Comment (nkwangjun) [reply hack/lib/common.sh]: Yes, this script is just for local test. You need to do start and pass the scheduler name like this way. Thanks

File: conf/conf.yaml (hunk 1 above)
Comment (nkwangjun) [reply conf/conf.yaml:6]: modified in  #146

File: conf/site_resources.json (hunk 2 above)
Comment (nkwangjun) [reply conf/site_resources.json:6]: modified in  #146

File: conf/edgecloud_policy.yaml (hunk 3 above)
Comment (nkwangjun) [reply conf/edgecloud_policy.yaml:14]: modified in  #146

File: docs/design-proposals/global-scheduler/GlobalSchedulerDemo.md (hunk 5 above)
Comment (nkwangjun) [reply docs/design-proposals/global-scheduler/GlobalSchedulerDemo.md]: modified in  #146

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 6 above)
Comment (nkwangjun) [reply cmd/kube-controller-manager/app/controllermanager.go]: modified in  #146

**Comment by nkwangjun** [comment]
All reivew comment is modified in  #146

**Pull request closed without merging**